### ISSUE TITLE
🚀 data cleaning interface

### ISSUE THREAD

**Issue by lukaszkruk** [kobotoolbox/kpi#2888]
## Description

Okay, so I think this is a reaaaly long shot, but I thought perhaps worth the few minutes of typing it up. I'm quietly hopeful more people think this would be useful and there's enough interest that it could be considered.

I think it is very common for small data collection projects to finish data collection, export a file, and then do some basic data cleaning locally - for example in Excel by looking at data collection mistakes that were not or could not be prevented through form logic constraints. This is more problematic for data collection efforts that last weeks or months - you want to have the ability to do frequent cleaning to follow up on mistakes, but downloading daily exports and later merging them into one file, and repeating the same cleaning steps over and over, gets tedious and error-prone.

I think it would be amazing if Kobo had a rules-based data cleaning interface, meaning a feature where you can select all submissions matching certain criteria, and then apply a correction to them. And then have those rules operate on all future submissions that meet the criteria. And then have the ability to get this cleaned data as an export, as well as power the autogenerated charts. And also keep the raw data, because it's of course always important to preserve the master dataset.

Basically, https://openrefine.org/ running within the Kobo web interface. 

I'll stop here for this post not to get too long, but happy to explain, justify, clarify further if helpful. Thanks in advance for any responses!

**Comment by joshuaberetta** [comment]
Hi @lukaszkruk, thank you for your great suggestion — clearly, some others are also interested in a feature like this. Unfortunately, we're unlikely to implement a big feature like this any time soon due to other obligations and higher priorities. 

In the meantime, we are planning to develop some [Jupyter Notebook](https://jupyter.org/) tutorials that could achieve this behavior outside of the Kobo web interface. It wouldn't be as slick of an interface as OpenRefine, but it will allow for automation once you have things set up.

If you're interested in this as a viable short-term alternative, it would be great to develop a Notebook with you that might solve your particular needs and then serve as a model for others to follow 💪

**Comment by lukaszkruk** [comment]
Hi @joshuaberetta, thanks for taking the time to review the suggestion and for your response!

Jupyter, as well as other data processing/analysis environments, are of course a possible solution. Colleagues at [REACH Initiative](https://www.reach-initiative.org/who-we-are/) and myself tend to use R, which works quite well. The problem with this approach for us, however, is that it tends to force the data to live on one (or a few at best) workstations, from which it then has to be sent further or published for use in web apps. Clumsy.

If you think this could be helpful, I could search for some example of our R scripts connecting to the Kobo API and then cleaning and exporting the data.

I recognise of course this is a big ask - but I don't really know how big. Your phrasing suggests that it is at least within the realm of possibility in the longer term. I'm wondering if it would be worth it to discuss if there are any resources we could contribute to making this more likely to happen?

**Comment by joshuaberetta** [comment]
@lukaszkruk, I understand that frustration with local cleaning. You could perhaps set up an automated cleaning approach with your current R scripts on a server and either pull from the API periodically or [register a REST service](https://support.kobotoolbox.org/rest_services.html) from your form. You can then populate a separate database with the clean data. The master data set will remain untouched on Kobo's servers (or yours if you are self-hosting). A bit of a hack, but should meet the underlying need.

If you can share those R scripts, that would be great 👌 

Out of interest, we are currently integrating some basic bulk editing / data cleaning into kpi at the moment and should be available soon. You can follow the progress of that over here: #2760, #2929, #2920, and #2941. Note that making changes through the new API endpoint or UI will update the data in the database.

With regards to contributing, you are of course welcome to submit a PR and we can decide whether or not to integrate it. You can find helpful resources for getting a development environment started [here](https://github.com/kobotoolbox/kobo-install) and [here](https://github.com/kobotoolbox/kpi/wiki/Frontend-Development-Troubleshooting). You are also welcome to join our "Developer Community" Zulip stream. We could also perhaps chat further about this with the Kobo and REACH teams to see if we have capacity and resources to work on this in the future.

**Comment by tinok** [comment]
@lukaszkruk Do you have some examples of the kinds of rules you would apply? Openrefine is great because it's very flexible, but I'm wondering if we could extend the updating feature we're working on with a more limited set of rules. 

The other option is to apply rules externally (e.g. through your R scripts) and then update the data on the KoBo server via the API. A script like https://github.com/kobotoolbox/xls-import could be used to send the data back and update records (using the original UUID and setting it as deprecated UUID). 

It seems that with either of these options you would want some way to easily go back and see the original unedited dataset? Or just access snapshots of individual submissions?
